xla/backends/cpu/runtime/onednn_thunk.cc

```diff
@@ -0,0 +1,150 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#if defined(INTEL_MKL) && defined(ENABLE_ONEDNN_V3)
+#include "xla/backends/cpu/runtime/onednn_thunk.h"
+
+#include <cstdint>
+#include <memory>
+#include <utility>
+
+#include "absl/memory/memory.h"
+#include "absl/status/statusor.h"
+#include "tsl/platform/statusor.h"
+#include "xla/backends/cpu/runtime/thunk.h"
+#include "xla/runtime/buffer_use.h"
+#include "xla/service/buffer_assignment.h"
+#include "xla/service/cpu/onednn_convolution.h"
+#include "xla/service/cpu/onednn_layer_norm.h"
+#include "xla/service/cpu/onednn_matmul.h"
+#include "xla/service/cpu/onednn_memory_util.h"
+#include "xla/service/cpu/onednn_softmax.h"
+#include "xla/stream_executor/device_memory.h"
+#include "xla/tsl/concurrency/async_value_ref.h"
+
+namespace xla::cpu {
+
+static Thunk::Kind DetermineThunkKind(const std::string& target) {
+  if (target == "__onednn$convolution") {
+    return Thunk::Kind::kOneDnnConvolution;
+  } else if (target == "__onednn$layernorm") {
+    return Thunk::Kind::kOneDnnLayerNorm;
+  } else if (target == "__onednn$matmul") {
+    return Thunk::Kind::kOneDnnMatMul;
+  } else if (target == "__onednn$matmul_reorder") {
+    return Thunk::Kind::kOneDnnMatMulReorder;
+  } else if (target == "__onednn$softmax") {
+    return Thunk::Kind::kOneDnnSoftmax;
+  } else {
+    LOG(FATAL) << "Unsupported OneDNN target: " << target;
+  }
+}
+
+OneDnnThunk::OneDnnThunk(const std::string& custom_call_target, Info info,
+                         OpBuffers buffers, absl::string_view config)
+    : target_(custom_call_target),
+      Thunk(DetermineThunkKind(custom_call_target), std::move(info)),
+      op_buffers_(buffers),
+      config_(config) {}
+
+absl::StatusOr<std::unique_ptr<OneDnnThunk>> OneDnnThunk::Create(
+    const std::string& custom_call_target, Info info, OpBuffers buffers,
+    absl::string_view config) {
+  return absl::WrapUnique(new OneDnnThunk(custom_call_target, std::move(info),
+                                          std::move(buffers), config));
+}
+
+tsl::AsyncValueRef<Thunk::ExecuteEvent> OneDnnThunk::Execute(
+    const ExecuteParams& params) {
+  std::vector<void*> args;
+  // Prepare args
+  // 1: nargs
+  // 2: ExecutableRunOptions
+  // 3: Config
+  // 4...: Operands
+
+  // nargs
+  int64_t nargs_offset = 3;
+  int64_t num_operands = op_buffers_.arguments_shapes.size();
+  int64_t num_args = nargs_offset + num_operands;
+  args.push_back(&num_args);
+
+  // ExecutableRunOptions
+  ExecutableRunOptions run_options;
+  run_options.set_intra_op_thread_pool(params.intra_op_threadpool);
+  args.push_back(&run_options);
+
+  // Config
+  args.push_back(config_.data());
```

Comment:
Future performance optimization is to refactor `_xla_cpu_runtime_OneDnnMatMul`, etc. to take config proto instead of serialized string (to save some deserialization time when the thunk is run).

(No actions required for this PR.)